File: src/main/java/me/oddlyoko/odoo/modules/models/ModuleDescriptor.java
package me.oddlyoko.odoo.modules.models;

import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import com.jetbrains.python.psi.PyDictLiteralExpression;
import com.jetbrains.python.psi.PyExpression;
import com.jetbrains.python.psi.PySequenceExpression;
import com.jetbrains.python.psi.PyStringLiteralExpression;
import com.jetbrains.python.psi.PyUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * A description of a module.<br />
 * Data is taken from manifest file
 */
public record ModuleDescriptor(PsiDirectory directory, PsiFile manifest, String id, String name, String version,
                               String description, List<String> author, String website, List<String> depends,
                               List<String> data, List<String> demo, String license) {


    @Override
    public String toString() {
        return id;
    }

    public static ModuleDescriptor parseFile(@NotNull PsiFile manifest) {
        PsiDirectory module = manifest.getParent();
        if (module == null)
            return null;
        PyDictLiteralExpression dict = PsiTreeUtil.findChildOfType(manifest, PyDictLiteralExpression.class);
        if (dict == null)
            return null;
        String id = module.getName();
        String name = "";
        String version = "";
        String description = "";
        List<String> authors = new ArrayList<>();
        String website = "";
        List<String> depends = new ArrayList<>();
        List<String> data = new ArrayList<>();
        List<String> demo = new ArrayList<>();
        String license = "";
        Map<String, PyExpression> map = PyUtil.dictValue(dict);
        for (Map.Entry<String, PyExpression> entry : map.entrySet()) {
            String key = entry.getKey();
            PyExpression value = entry.getValue();
            switch (key) {
                case "name" -> {
                    if (value instanceof PyStringLiteralExpression stringExpression)
                        name = stringExpression.getStringValue();
                }
                case "version" -> {
                    if (value instanceof PyStringLiteralExpression stringExpression)
                        version = stringExpression.getStringValue();
                }
                case "description" -> {
                    if (value instanceof PyStringLiteralExpression stringExpression)
                        description = stringExpression.getStringValue();
                }
                case "author" -> {
                    if (value instanceof PyStringLiteralExpression stringExpression)
                        authors = List.of(stringExpression.getStringValue());
                    else if (value instanceof PySequenceExpression sequenceExpression)
                        authors = sequenceToList(sequenceExpression);
                }
                case "website" -> {
                    if (value instanceof PyStringLiteralExpression stringExpression)
                        website = stringExpression.getStringValue();
                }
                case "depends" -> {
                    if (value instanceof PySequenceExpression sequenceExpression)
                        depends = sequenceToList(sequenceExpression);
                }
                case "data" -> {
                    if (value instanceof PySequenceExpression sequenceExpression)
                        data = sequenceToList(sequenceExpression);
                }
                case "demo" -> {
                    if (value instanceof PySequenceExpression sequenceExpression)
                        demo = sequenceToList(sequenceExpression);
                }
                case "license" -> {
                    if (value instanceof PyStringLiteralExpression stringExpression)
                        license = stringExpression.getStringValue();
                }
            }
        }
        return new ModuleDescriptor(module,
                manifest,
                id,
                name,
                version,
                description,
                authors,
                website,
                depends,
                data,
                demo,
                license);
    }

    private static List<String> sequenceToList(PySequenceExpression pySequence) {
        List<String> lst = PyUtil.strListValue(pySequence);
        // Don't ask me why PyUtil.strListValue returns null and not an empty List
        if (lst == null)
            return List.of();
        else
            return Collections.unmodifiableList(lst);
    }
}

File: src/main/kotlin/me/oddlyoko/odoo/modules/models/ModuleDescriptor.kt
package me.oddlyoko.odoo.modules.models

import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import com.jetbrains.python.psi.PyDictLiteralExpression
import com.jetbrains.python.psi.PySequenceExpression
import com.jetbrains.python.psi.PyStringLiteralExpression
import com.jetbrains.python.psi.PyUtil

data class ModuleDescriptor(val directory: PsiDirectory, val manifest: PsiFile, val name: String, val id: String,
                            val version: String, val description: String, val authors: List<String>,
                            val website: String, val depends: List<String>, val data: List<String>,
                            val demo: List<String>, val license: String) {

    override fun toString() = id

    companion object {
        fun fromFile(manifest: PsiFile): ModuleDescriptor? {
            val module: PsiDirectory = manifest.parent ?: return null
            val dict = PsiTreeUtil.findChildOfType(manifest, PyDictLiteralExpression::class.java) ?: return null
            val id = module.name
            var name = ""
            var version = ""
            var description = ""
            var authors = listOf<String>()
            var website = ""
            var depends = listOf<String>()
            var data = listOf<String>()
            var demo = listOf<String>()
            var license = ""

            PyUtil.dictValue(dict).forEach { (key, value) -> run {
                when (key) {
                    "name" -> {
                        if (value is PyStringLiteralExpression)
                            name = value.stringValue
                    }
                    "version" -> {
                        if (value is PyStringLiteralExpression)
                            version = value.stringValue
                    }
                    "description" -> {
                        if (value is PyStringLiteralExpression)
                            description = value.stringValue
                    }
                    "author" -> {
                        when (value) {
                            is PyStringLiteralExpression -> authors = listOf(value.stringValue)
                            is PySequenceExpression -> authors = seqToList(value)
                        }
                    }
                    "website" -> {
                        if (value is PyStringLiteralExpression)
                            website = value.stringValue
                    }
                    "depends" -> {
                        if (value is PySequenceExpression)
                            depends = seqToList(value)
                    }
                    "data" -> {
                        if (value is PySequenceExpression)
                            data = seqToList(value)
                    }
                    "demo" -> {
                        if (value is PySequenceExpression)
                            demo = seqToList(value)
                    }
                    "license" -> {
                        if (value is PyStringLiteralExpression)
                            license = value.stringValue
                    }
                }
            } }

            return ModuleDescriptor(module, manifest, id, version, description, name, authors, website, depends, data, demo, license)
        }

        private fun seqToList(seq: PySequenceExpression): List<String> = PyUtil.strListValue(seq) ?: listOf()
    }
}
